[IBD patch 4] Deprioritise slower downloaders

ptschip opened the pull request:
Here we adjust the max blocks in flight per peer based on it's historical download response times.  This way we ask for more blocks from faster peers.

sickpig committed:
Use Logging namespace to access LogAcceptCategory

LogAcceptCategory and the enum used to define available logging
category is defined inside the Logging namespace. If used directly
outside of util.{h,cpp} reference to such namespace is needed.

This commit fix issue #950

ptschip committed:
Update the block availaibility for all nodes during IBD

One of the problems with IBD is that we end up downloading
blocks from the same peer. This is because after downloading
all the headers at startup the block availability is only
being updated for that peer. We need to assume that all connected
or newly connected peers have the blocks that we need but we
only do this during IBD or when IsInitialBlockDownload() is true.

ptschip committed:
Increase the block download window to 1024

This helps to keep IBD moving forward when we have a peer
that may be serving us blocks more slowly.  And now that
blockvailability is being updated correctly we are getting blocks
from all connected peers so the chance of having a slow peer
is now much higher.

ptschip committed:
Reset the single peer request mode age to 24 hours

One week isn't needed anymore because the new DAA is working
well and we're not falling behind in terms of blocks mined
per day.

ptschip committed:
Remove the check for node ping in the request manager

This feature is causing occasional hangups during IBD and
is not really effective in selecting peers especially during
IBD. Since it is not used when the chain is sync'd then there
is no need for this feature to remain which only adds another
level of complexity to IBD. KISS can be applied here for a better
IBD experience.

ptschip committed:
Only update xthin stats when IsChainNearlySyncd()

There is no need to lock vNodes and check all peers for
thinblocks in flight unless we are IsChainNearlySyncd() because
we would not have asked for any xthins if !IsChainNearlySyncd().

This is an expensive operation and which makes a performance hit
during IBD.

ptschip committed:
Disconnect non NETWORK nodes during initial sync

We only want to have nodes we can download the full
blockchain from, connected to us while we do our initial sync.

ptschip committed:
Increase the default max outbound connections to 16

Having a few more outbound connections can help especially
during the process of IBD.

ptschip committed:
During IBD, batch the block reqeusts if possible.

By batching the blocks together we save a little bandwidth
and time in requesting blocks.

This method can then be applied to batching transactions which
should make a very good improvement during periods where transaction
rates are high.

ptschip committed:
Replace missing cs_main lock

ptschip committed:
fix formatting

ptschip committed:
Calculate MaxBlocksInTransitPerPeer, but on an individual node basis

Before we were using the overall resonse times regardless of
which peer they came from. This made the selection of how many
blocks to donwload at one time not responsive to faster or
slower nodes.  By tracking response times on a per node basis
we can deprioritize slower downloaders and get more blocks
from the faster nodes.

In HEADERS net processing, remove MAX_BLOCKS_IN_TRANSIT_PER_PEER

  This has been replaced with pnode->nMaxBlocksInTransitPerPeer

use atomic store, load, fetch_add

ptschip committed:
Fix locking order issue with cs_objectDownloader and cs_vNodes

cs_objectDownloader must be locked first to prevent a possible
deaklock.

ptschip committed:
Fix formatting

sickpig commented on src/main.cpp:
```diff
@@ -7530,30 +7530,35 @@ bool SendMessages(CNode *pto)
+                if (!IsInitialBlockDownload())
+                    requester.AskFor(vGetBlocks, pto);
+                else
+                    requester.AskForDuringIBD(vGetBlocks, pto);
```
wonder if we could have only one `AskFor` to which we pass another bool par to indicate if we are in the IBD phase or not

ptschip replied:
@sickpig I made your requested changed but I've since changed it back.  While in general you are right to suggest this, in this particular case it doesn't work very well.  The problem is that the AskFor() method is also used for getting transactions also; this then makes the code somewhat confusing when I inject code specific for IBD, but in addition we have to then make AskFor() recursive.  I found after I wrote the fix you requested that the code would be very hard to read and make sense for someone unfamiliar with the code, of and coupled with a recursive call that could potentially be a problem if someone in the future updated the code and made an unintended change.  So I thought better to make the code readable and more safe and leave it as it is with the AskForDuringIBD().

sickpig replied:
fair enough

sickpig commented on src/unlimited.cpp:
```diff
@@ -1246,7 +1247,18 @@ void IsInitialBlockDownloadInit()
+            // see if in fact we have the full set of headers; It was possible that during the initial header download
+            // we didn't get everything, either because the peer dropped off or because they fed us an imcomplete set.
+            nMaxHeaderReqsDuringIBD += 3;
+            LOGA("IBD complete - increase max header reqquests to check if we have the full set of headers\n");
```
s/reqquests/request

awemany commented on src/net.cpp:
```diff
@@ -1745,7 +1745,8 @@ void ThreadOpenConnections()
         int nThinBlockCapable = 0;
         set<vector<unsigned char> > setConnected;
-        CNode *ptemp = nullptr;
+        CNode *ptemp1 = nullptr;
```
ptemp1 and ptemp2 are pretty generic variable names, making this hard to follow :)

ptschip committed:
Don't favor XTHIN nodes when sending messages

This has a negative effect on IBD

ptschip committed:
Don't ask for the same blocks twice when doing IBD

If we've already asked for a block, we don't have to ask for it
again in FindNextBlocksToDownload(), instead we can rely o the request
manager handle potential re-requests.

ptschip commented:
thank you ... done.

On 14/02/2018 6:39 AM, Andrea Suisani wrote:
>
> *@sickpig* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/main.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/970#discussion_r168191407>:
>
> >                  }
>              }
> +            if (!vGetBlocks.empty())
> +            {
> +                if (!IsInitialBlockDownload())
> +                    requester.AskFor(vGetBlocks, pto);
> +                else
> +                    requester.AskForDuringIBD(vGetBlocks, pto);
>
> wonder if we could have only one |AskFor| to which we pass another
> bool par to indicate if we are in the IBD phase or not
>
> ------------------------------------------------------------------------
>
> In src/unlimited.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/970#discussion_r168193119>:
>
> > @@ -1246,7 +1247,18 @@ void IsInitialBlockDownloadInit()
>          (chainActive.Height() < pindexBestHeader->nHeight - 24 * 6 ||
>              std::max(chainActive.Tip()->GetBlockTime(), pindexBestHeader->GetBlockTime()) < GetTime() - nMaxTipAge);
>      if (!state)
> +    {
> +        if (!fInitialSyncComplete)
> +        {
> +            // As we cross the boundary for the first time from IsInitialBlockDownload to !IsInitialBlockDownload, we
> +            // want to increment the max header reqs.  What this does is allow us to go back and check more peers to
> +            // see if in fact we have the full set of headers; It was possible that during the initial header download
> +            // we didn't get everything, either because the peer dropped off or because they fed us an imcomplete set.
> +            nMaxHeaderReqsDuringIBD += 3;
> +            LOGA("IBD complete - increase max header reqquests to check if we have the full set of headers\n");
>
> s/reqquests/request
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/970#pullrequestreview-96517219>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0H74C2j59Jozo844AcwLOsNTkQGnks5tUvA6gaJpZM4SER7T>.
>

ptschip commented:
good idea, having trouble myself following it  ... done.

On 14/02/2018 7:12 AM, awemany wrote:
>
> *@awemany* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/net.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/970#discussion_r168203909>:
>
> > @@ -1745,7 +1745,8 @@ void ThreadOpenConnections()
>          int nOutbound = 0;
>          int nThinBlockCapable = 0;
>          set<vector<unsigned char> > setConnected;
> -        CNode *ptemp = nullptr;
> +        CNode *ptemp1 = nullptr;
>
> ptemp1 and ptemp2 are pretty generic variable names, making this hard
> to follow :)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/970#pullrequestreview-96532283>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0Aph2xxaf2jwxL_H82QMqc-3UzpNks5tUvf1gaJpZM4SER7T>.
>

ptschip committed:
Use more descriptive variable names for nodes we are disconnecting

ptemp1 and ptemp2 are getting hard to follow. We'll say what they are:

pNonXthinNode and pNonNodeNetwork

ptschip committed:
Take the cs_objDownloader lock earlier when requesting multiple objects

This prevents anyone else from asking for these same objects before
we've notified the request manager of their existence. This could happen
for in stance during IBD where we look for the next blocks to download
but before we've notified the request manager of them all we then
go back and possibly request them again.

Also keeping the lock here allows the request manager to prepare
a better batch request of blocks and thereby keep a better order
of block requests.

sickpig commented:
I've tested the patch  bootstrapping a full nodes multiple times from scratch (modulo the last commit related to `cs_objDownloader` lock) and the IBD time is on par with Core taking into consideration the difference in size of the 2 block chains. Before these set of PRs BU IBD time was significantly higher.

ptschip commented:
@gandrewstone @sickpig 

Test Results:

Completed 2 full syncs running on  4core i7 8th gen, 16GB RAM storing to SSD, with a wired 150Mbit internet connection and 10GB dbcache.

Ran two tests, 1 a BU node running IBD patch 4, the other ABC node v16.2

BU took 3hr 4  min, ABC 3 hr 23min.  The results are interesting in that they point to even further possible improvements in BU.  The ABC node , although slower overall, really blew away the BU node in the first part of the block chain...after 20 mins, BU was at 245K blocks wheres ABC was already at 265K...however ABC lost all that ground and some in the latter half of the sync.  My observation/analysis shows that it was PV that allowed BU to make up the ground.  The ABC node suffered in the latter part of the sync and could be observed having very large swings in bandwidth use, wheras the BU node was smooth and steady.  I believe this is the result of PV which allows the BU node to ask for and download blocks while it is currently connecting the next block.  So why did ABC do so well in the beginning?  I think it has something to do with the script cache they have which we haven't implemented yet.  I think once we do that  we should be able to shave off another 10 to 20 mins of sync time...once these set of changes are ready and merged I'll work on bringing over the script caching.

ptschip committed:
Add/release node refs when making batch requests

We need to track the node refs correctly so we don't get disconnected
before we're done our work.

ptschip committed:
Call MarkBlocksAsInFlight() before asking for blocks

This ensures you don't receive any blocks back before you mark the block
in flight as could happen on regtest.

ptschip commented:
@gandrewstone I got the last bug that was causing an intermittent problem.  It was MarkBlocksAsInFlight() which has to be called "BEFORE" we send the block requests , or else we can end up getting blocks back before the block is marked in flight causing a crash.   I think this code is ready for final review.

ptschip commented:
looks like there is still a bug in here somewhere....intermittent python test failures but only happen on Travis which makes it difficult to debug.

sickpig commented:
@ptschip I rekick the failing travis job to see if the failure is spurious.

ptschip commented:
I think this is ready, re-worked this PR and did some cleanup...no issues with travis runs.

Ran a full sync on  e9fb183  forgot to plugin my ethernet cable so this one was over wireless but still was very good...only 4 minutes slower than the previous run, coming in at just under 3hrs and 8 minutes.  Previous run on a wired connections was 3 hrs 4 mins.

ptschip committed:
Fixes #941: Rotate vNodes by one peer every 60 seconds when sending messages

We do this only during IBD and this has the effect of distributing
the load more evenly between the peers. Previously, because we are
using PV, the very first peer to connect would always be favored and
we would end up downloading a disproportionate amount of blocks from
that peer during IBD.  However, we still will download more from
some peers based on their download response time performance.

ptschip committed:
Small net optimization when sending messages

Just copy the entire vector rather than bothering to allocate
memory and then push_back for each entry as we iteration through
vNodes.

Make sure to LOCK vNodes then add the ref

ptschip commented:
@gandrewstone @sickpig I fixed one of these mysterious terminations we get on travis...I thought it was something related to this PR but it turns out not...Still the fix is just a workaround...I think the problem has something to do with RPC connections not getting released right away, so by sleeping for  a second before doing join_network (after doing a getchaintips) seems to prevent the script from hanging.

sickpig commented:
I tested the PR with the last changes checked in e3b6640 and I got even better performance than the latest test I did last week. The IBD improve significantly going from ~10 hrs to ~7hrs (using `-dbcache=10000`).

sickpig commented on qa/rpc-tests/getchaintips.py:
```diff
@@ -9,7 +9,7 @@
 # This gives us two tips, verify that it works.
 
 from test_framework.test_framework import BitcoinTestFramework
-from test_framework.util import assert_equal
```
nit: why import everything?

ptschip replied:
because i was feeling momentarily lazy...i'll modify it and just use "time"

ptschip commented:
One last spurious failure to go...this is the one that happens in excessive.py but it seems to get triggered more frequently in this patch...problem is , this is another one that only happens on travis...yuk,

ptschip commented:
Still getting random failures...hmmm...

sickpig commented:
@ptschip are you experiencing random failure on you red machine or on travis?

sickpig commented on src/requestManager.cpp:
```diff
@@ -164,13 +164,43 @@ void CRequestManager::AskFor(const CInv &obj, CNode *from, unsigned int priority
     }
 }
 
+void CRequestManager::AskForDuringIBD(const std::vector<CInv> &objArray, CNode *from, unsigned int priority)
```
why not using uint16/32_t as priority type? (if we change here we need to change it also on `AskFor`)

ptschip commented:
@sickpig at first only travis, but now i've been able to replicate  (at times) on my own local machine.  Something in the last IBD4 patch that has triggered it but it's proving exceedingly difficult to track down.

ptschip commented:
putting this in WIP while I continue to investigate the cause of the random python test failures for this PR... the previous PR's  IBD1, IBD2 and IBD3 don't have the random failures, only IBD4 has them.

ptschip committed:
Fix getchaintips.py spurious failures

The hang results from a  PV bug which is caused by a condition where
threads that are waiting for validation are forced to quit during chain
tip rollback.  We don't in fact have to force any threads to quit since
any validation threads on the current chain will fail automatically
once we disconnect the tip.  And then the waiting block validation
thread for the new chain can still continue without being inadvertanly
forced to quit.

ptschip commented:
Finally resolved the spurious failure in getchaintips.py

ptschip committed:
Do not to use an incorrect or invalid state

Use the node id from the iterator rather than looking up state
by the node we think we just passed in. Just in case they don't match
for some reason.

Also DbgAssert if the state is NULL and return false if we assert.

gandrewstone commented on src/main.cpp:
```diff
@@ -294,8 +291,11 @@ void FinalizeNode(NodeId nodeid)
-    BOOST_FOREACH (const QueuedBlock &entry, state->vBlocksInFlight)
+    for (const QueuedBlock &entry : state->vBlocksInFlight)
     {
+        printf("erasing map mapblocksinflight entries\n");
```
remove

ptschip replied:
done

gandrewstone commented on src/main.cpp:
```diff
@@ -313,67 +313,79 @@ void FinalizeNode(NodeId nodeid)
-        LOG(THIN, "BLOCK_DOWNLOAD_WINDOW is %d MAX_BLOCKS_IN_TRANSIT_PER_PEER is %d\n", BLOCK_DOWNLOAD_WINDOW,
-            MAX_BLOCKS_IN_TRANSIT_PER_PEER);
+        LOG(THIN, "BLOCK_DOWNLOAD_WINDOW is %d nMaxBlocksInTransit is %d\n", BLOCK_DOWNLOAD_WINDOW,
+            pnode->nMaxBlocksInTransit.load());
```
is there a better category for this?  Maybe BLK | THIN

ptschip replied:
done...added BLK

gandrewstone commented on src/requestManager.cpp:
```diff
@@ -164,13 +164,43 @@ void CRequestManager::AskFor(const CInv &obj, CNode *from, unsigned int priority
+        if (pnode == from)
+            continue;
+
+        AskFor(objArray, pnode, priority);
```
it looks here like you are asking for every block from every connected node?

ptschip replied:
yes, we have to do this here.  We need to add a block source for every peer otherwise we can't do re-requests.  But what I should do is (1) only add sources for NETWORK nodes and then (2) perhaps filter for peers that have block availablility , if I can actually do that here, and if I can get the correct block availability ... i'll work on this along with getting the correct block availability for each peer rather than assuming they all have the blocks.

sickpig replied:
was wondering, if a node as NODE_NETWORK set and its current height (CH) is higher than our current best chain tip (CT), what is the reason why it wouldn't be able to provide to us any blocks with height H belonging to the (CT, CH]?

gandrewstone commented on src/requestManager.cpp:
```diff
@@ -587,6 +617,31 @@ void CRequestManager::SendRequests()
+            }
+            iter.first->PushMessage(NetMsgType::GETDATA, iter.second);
+            ENTER_CRITICAL_SECTION(cs_objDownloader);
+            LOG(REQ, "Sent batched rqst with %d blocks to node %s\n", iter.second.size(), iter.first->GetLogName());
```
rqst -> request

ptschip replied:
done

ptschip committed:
Small edit to logprint

Add category BLK to logprint

remove printf

sickpig commented:
travis failure is due to a formatting issue not a tech one

```c++
--- main.cpp
+++ main.cpp_formatted
@@ -362,7 +362,7 @@
                 pnode->nMaxBlocksInTransit.store(16);
             }
 
-            LOG(THIN|BLK, "Average block response time is %.2f seconds\n", pnode->nAvgBlkResponseTime);
+            LOG(THIN | BLK, "Average block response time is %.2f seconds\n", pnode->nAvgBlkResponseTime);
         }
 
         // if there are no blocks in flight then ask for a few more blocks
@@ -377,7 +377,7 @@
         {
             BLOCK_DOWNLOAD_WINDOW = blockDownloadWindow.value;
         }
-        LOG(THIN|BLK, "BLOCK_DOWNLOAD_WINDOW is %d nMaxBlocksInTransit is %d\n", BLOCK_DOWNLOAD_WINDOW,
+        LOG(THIN | BLK, "BLOCK_DOWNLOAD_WINDOW is %d nMaxBlocksInTransit is %d\n", BLOCK_DOWNLOAD_WINDOW,
             pnode->nMaxBlocksInTransit.load());
 
         if (IsChainNearlySyncd())
```

ptschip committed:
During IBD only request blocks from peers that have block availability

Request from peers that are NODE_NETWORK and have demonstrated that
they have the blocks that we need.

Also move ProcessBlockAvailablity and UpdateBlockAvailabilty to
RequestManager.cpp

sickpig commented:
>   So that's why I added the code to assume that every peer we connected to has the blocks we need.  

during IBD I think this is a fair assumption, given that the peer is signaling NODE_NETWORK., i.e. not a pruned node. We could evaluate if the peer is a "good" peer in terms of bandwidth, but I would expect that any given nodes that:

- signals NODE_NETWORK (not pruned)
- has a chaintip higher than our own (or that it's fully synced) 

even if didn't provide us the header of the block at hand should have the block data. 

> However, as Andrew got me thinking about it, I think we can improve that here by first only doing AskFor() from peers that have block availability (even though we assume they do anyway) and then while we're downloading headers from those 3 initial peers we can also request a header occasionally from every other peer to correctly update everybody block availability.

fair enough.

ptschip committed:
Move FindNextBlocksToDownload() to the requestManager.cpp

Also change all NULL's to nullptr.

ptschip committed:
Move MarkBlockAsInFlight() and MarkBlockAsReceived() to requestManager.cpp

This ia a Move only.  Alhtough there are edits in order to
bring these into the requestManager class structure there
are no logic changes.

C++ nullptr replaces NULL's

ptschip commented:
This last series of commits was really just intended to check the correct block availability in AskForDuringIBD()  however I had to move over ProcessBlockAvailaibility and UpdateBlockAvailability to the request manager...then I thought , why not move over FindNextBlocksToDownload() as we would have the IBD code all in one place.  Then I got a little rambunctious and decided to finish the job and bring over MarblocksAsInFlight and MarkBlockAsReceived() as that would have all the block requesting code finally in the request manager which I've been wanting to do for a year now.  There are no logic changes to any of the moves but it but If it's too much I can back so it out for now, but I still need Processblockavailability/Updateblockavailability anyway as they were giving issues.

ptschip commented:
I have one more item to work on which is getting the correct blockavailability from each peer as we're requesting the initial set of headers during sync.

ptschip committed:
Update block availablity during the process of downloading headers

During initial sync we have to download a set of all headers but
we also need to update the block availabiity for each connected
peer so that when the request manager starts downloading blocks
it can have the correct list of peers that have available blocks
to download from, rather than just assuming every peer has all
blocks.

ptschip commented:
I think this last commit is good...doing some testing now.

ptschip commented:
Testing looks good...sync is the same as before 200K + blocks in the first 10 minutes of sync.  I verified we're downloading from all network peers so it looks like the block availability is updating correctly.  I think this is good for final review...

sickpig commented:
Going to start looking at the code while doing a test IBD using the latest version of the PR

gandrewstone commented on src/main.cpp:
```diff
@@ -6398,6 +6098,26 @@ bool ProcessMessage(CNode *pfrom, std::string strCommand, CDataStream &vRecv, in
+                {
+                    if (!pnode->fClient && pnode != pfrom)
+                    {
+                        // We only want one single header so we pass a null for CBlockLocator.
```
If the block has already given us its best headers, this PushMessage might be unnecessary.  I am also worried about an infinite request loop.  right now this can't happen because PushMessage only asks for a single header, but this code only triggers if nCount == MAX_HEADERS_RESULTS.  But the fact that this prevents an request loop is a secondary effect that might be missed by a future developer.  We can solve both of these by checking that the node doesn't already have a better header here.  I see that check in UpdateBlockAvailability:

if (state->pindexBestKnownBlock == NULL || it->second->nChainWork >= state->pindexBestKnownBlock->nChainWork)
            state->pindexBestKnownBlock = it->second;

Does it make sense to do a similar check here and not send a GETHEADERS if our best block already has greater work than this one?

ptschip replied:
hmmm, yes i think you're right, but the problem is cs_main again and locking order...will have to create a copy of vnodes, update refs, release ref's etc...it really bugs me that state is dependent on cs_main...shouldn't have to be...i'll look at that at another time, but for now i'll work on your requested change.

ptschip replied:
@gandrewstone ok, done...

gandrewstone commented on src/requestManager.cpp:
```diff
@@ -587,6 +667,31 @@ void CRequestManager::SendRequests()
+    {
+        for (auto iter : mapBatchBlockRequests)
+        {
+            LEAVE_CRITICAL_SECTION(cs_objDownloader);
```
we should be able to put this and the ENTER outside of the for loop

ptschip replied:
thanks, ok done...also moved up cs_main...we only need to take cs_main once as well.

gandrewstone commented on src/requestManager.cpp:
```diff
@@ -587,6 +667,31 @@ void CRequestManager::SendRequests()
+            LOG(REQ, "Sent batched request with %d blocks to node %s\n", iter.second.size(), iter.first->GetLogName());
+        }
+
+        LOCK(cs_vNodes);
```
This lock is not be necessary I think.  Since we have ownership of the node, it can't be removed from cs_vNodes

ptschip committed:
EXIT cs_objDownloader earlier.

We don't have to repeatedly take and release the locks.
Just do it once for all items in the entire batch request.

ptschip committed:
Only ask for headers to updateblockavailability if chain work is behind

When doing the initial request for headers and we're in the process
of also updating the block availability for each connected peer we
only need to ask for a header if any peer is actually behind in
terms of chain work. This prevents first of all requesting
headers we don't need but also any possible attack where we're being
fed an invalid group of headers which then causes us to request
large number of additional headers.
